dbms/src/Storages/Transaction/SSTReader.h

```diff
@@ -14,28 +14,119 @@
 
 #pragma once
 
+#include <Common/Logger.h>
 #include <Common/nocopyable.h>
 #include <Storages/Transaction/ProxyFFI.h>
 
 namespace DB
 {
+class SSTReader
+{
+public:
+    virtual bool remained() const = 0;
+    virtual BaseBuffView keyView() const = 0;
+    virtual BaseBuffView valueView() const = 0;
+    virtual void next() = 0;
+
+    virtual ~SSTReader(){};
+};
 
-struct SSTReader
+class MonoSSTReader : public SSTReader
 {
-    bool remained() const;
-    BaseBuffView key() const;
-    BaseBuffView value() const;
-    void next();
+public:
+    bool remained() const override;
+    BaseBuffView keyView() const override;
+    BaseBuffView valueView() const override;
+    void next() override;
 
-    DISALLOW_COPY_AND_MOVE(SSTReader);
-    SSTReader(const TiFlashRaftProxyHelper * proxy_helper_, SSTView view);
-    ~SSTReader();
+    DISALLOW_COPY_AND_MOVE(MonoSSTReader);
+    MonoSSTReader(const TiFlashRaftProxyHelper * proxy_helper_, SSTView view);
+    ~MonoSSTReader() override;
 
 private:
     const TiFlashRaftProxyHelper * proxy_helper;
     SSTReaderPtr inner;
     ColumnFamilyType type;
 };
 
+/// MultiSSTReader helps when there are multiple sst files in a column family.
+/// It is derived from virtual class SSTReader, so it can be holded in a SSTReaderPtr.
+/// It also maintains instance of `R` which is normaly SSTReader(and MockSSTReader in tests),
+/// to read a single sst file.
+/// An `Initer` function is need to create a instance of `R` from a instance of `E`,
+/// which is usually path of the SST file.
+/// We introduce `R` and `E` to make this class testable.
+template <typename R, typename E>
+class MultiSSTReader : public SSTReader
+{
+public:
+    using Initer = std::function<std::unique_ptr<R>(const TiFlashRaftProxyHelper *, E)>;
+
+    DISALLOW_COPY_AND_MOVE(MultiSSTReader);
+
+    // Whether the current key is valid.
+    bool remained() const override
+    {
+        return mono->remained();
+    }
+    BaseBuffView keyView() const override
+    {
+        return mono->keyView();
+    }
+    BaseBuffView valueView() const override
+    {
+        return mono->valueView();
+    }
+    void next() override
+    {
+        mono->next();
+        // If there are no remained keys, we try to switch to next mono reader.
+        this->maybe_next_reader();
+    }
+
+    // Switch to next mono reader if current is drained,
+    // and we have a next sst file to read.
+    void maybe_next_reader() const
+    {
+        if (!mono->remained())
+        {
+            current++;
+            if (current < args.size())
+            {
+                // We don't drop if mono is the last instance for safety,
+                // and it will be dropped as MultiSSTReader is dropped.
```

Comment:
What kind of safety do you want to have? For example, not crashing the TiFlash due to nullptr when there are `valueView()` after `next()` and drained or something else?

Reply:
We the mono hold is the last sst file, `MultiSSTReader` acts like a `MonoSSTReader`